### PR TITLE
InteractionService Feature: Passing CustomId matches into contexts

### PR DESCRIPTION
**Pull request by Cenngo**
This PR introduces a new interface called `IRouteMatchContainer`. This interface can be optionally implemented into Interaction Context types to allow the interaction service to inject the captured wild card values into an interaction context. Captured values are exposed to the user using a `IRouteSegmentMatch` collection and not as a `string[]` as a futureproofing measure. While this allows the users to access these values within preconditions, BeforeExecute, AfterExecute methods; ParameterPrecondition should be preferred to get and process individual parameter values.

### CODE REVIEW

**Comment by quinchs** [comment]
@Cenngo Whats the status on this?

**Comment by Cenngo** [comment]
> @Cenngo Whats the status on this?

To be honest i forgot about this after i started working on Discord-Net-Labs/Discord.Net-Labs#449 migration. Ill wrap this up and let you know.

**Comment by Cenngo** [comment]
@quinchs This PR is in working state but i dont know if this change is necessary or if this is the best way to implement this. would like to hear your thoughts on this

File: src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /// <summary>
+    ///     Represents a container for temporarily storing CustomId wild card matches of a command name.
+    /// </summary>
+    public interface IRouteMatchContainer
+    {
+        /// <summary>
+        ///     Gets the collection of the captured route segments.
+        /// </summary>
+        public IEnumerable<IRouteSegmentMatch> SegmentMatches { get; }
```
Comment (quinchs) [inline src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs]:
```suggestion
        IEnumerable<IRouteSegmentMatch> SegmentMatches { get; }
```

File: src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+        /// <summary>
+        ///     Sets the <see cref="SegmentMatches"/> propert
+        /// </summary>
+        /// <param name="segmentMatches">The collection of captured route segments.</param>
+        public void SetSegmentMatches(IEnumerable<IRouteSegmentMatch> segmentMatches);
```
Comment (quinchs) [inline src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs]:
```suggestion
        void SetSegmentMatches(IEnumerable<IRouteSegmentMatch> segmentMatches);
```

**Review by quinchs — CHANGES_REQUESTED** [review]

**Comment by quinchs** [comment]
cc @Rozen4334

**Review by quinchs — APPROVED** [review]

File: src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+using System.Collections.Generic;
+
+namespace Discord
+{
+    /// <summary>
+    ///     Represents a container for temporarily storing CustomId wild card matches of a command name.
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs]:
```suggestion
    ///     Represents a container for temporarily storing CustomId wild card matches of a component.
```

File: src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    /// </summary>
+    public interface IRouteMatchContainer
+    {
+        /// <summary>
+        ///     Gets the collection of the captured route segments.
+        /// </summary>
+        IEnumerable<IRouteSegmentMatch> SegmentMatches { get; }
+
+        /// <summary>
+        ///     Sets the <see cref="SegmentMatches"/> propert
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs]:
```suggestion
        ///     Sets the <see cref="SegmentMatches"/> property of this container.
```

File: src/Discord.Net.Core/Interactions/IRouteSegmentMatch.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+namespace Discord
+{
+    /// <summary>
+    ///     Represents an object for storing CustomId a wild card match.
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteSegmentMatch.cs]:
```suggestion
    ///     Represents an object for storing a CustomId wild card match.
```

File: src/Discord.Net.Core/Interactions/IRouteSegmentMatch.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    /// </summary>
+    public interface IRouteSegmentMatch
+    {
+        /// <summary>
+        ///     Gets the captured value.
+        /// </summary>
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteSegmentMatch.cs:10]:
```suggestion
        /// </summary>
        /// <returns>
        ///    The value of this wild card.
        /// </returns>
```

File: src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+namespace Discord
+{
+    /// <summary>
+    ///     Represents a container for temporarily storing CustomId wild card matches of a command name.
+    /// </summary>
+    public interface IRouteMatchContainer
+    {
+        /// <summary>
+        ///     Gets the collection of the captured route segments.
+        /// </summary>
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs:12]:
```suggestion
        /// </summary>
        /// <returns>
        ///    A collection of captured route segments.
        ///</returns>
```

File: src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Discord
+{
+    /// <summary>
+    ///     Represents a container for temporarily storing CustomId wild card matches of a command name.
+    /// </summary>
+    public interface IRouteMatchContainer
+    {
+        /// <summary>
+        ///     Gets the collection of the captured route segments.
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteMatchContainer.cs]:
```suggestion
        ///     Gets the collection of captured route segments in this container.
```

File: src/Discord.Net.Core/Interactions/IRouteSegmentMatch.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+namespace Discord
+{
+    /// <summary>
+    ///     Represents an object for storing CustomId a wild card match.
+    /// </summary>
+    public interface IRouteSegmentMatch
+    {
+        /// <summary>
+        ///     Gets the captured value.
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/IRouteSegmentMatch.cs]:
```suggestion
        ///     Gets the captured value of this wild card match.
```

File: src/Discord.Net.Core/Interactions/RouteSegmentMatch.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+namespace Discord
+{
+    /// <summary>
+    ///     Represents an object for storing CustomId wild card matches.
```
Comment (csmir) [inline src/Discord.Net.Core/Interactions/RouteSegmentMatch.cs]:
```suggestion
    ///     Represents an object for storing a CustomId wild card match.
```

**Review by csmir — CHANGES_REQUESTED** [review]

**Review by csmir — APPROVED** [review]
lgtm

**Pull request merged**